### PR TITLE
NIFI-7230: Created toInstant() and formatInstant() EL functions

### PR DESCRIPTION
**Pull request by Lehel44**
…e functions

Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

https://issues.apache.org/jira/browse/NIFI-7230

Added **toInstant** and **formatInstant** functions to EL to support nanosecond precision datetime.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [x] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] Have you verified that the full build is successful on JDK 8?
- [ ] Have you verified that the full build is successful on JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by Lehel44** [comment]
Probably will add precision argument to the toNumber() function soon.

File: nifi-docs/src/main/asciidoc/expression-language-guide.adoc [hunk 1]
```diff
@@ -2233,8 +2233,40 @@ In order to run the correct method, the parameter types must be correct. The Exp
...
 
 
+[[formatInstant]]
+[.function]
+=== formatInstant
+
+*Description*: [.description]#Formats a number or string as a date/time according to the format specified by the argument.
+The argument must be a String that is a valid Java DateTimeFormatter format. The Subject is expected to be a Number that
+represents the number of milliseconds since Midnight GMT on January 1, 1970 or a String with one of the following
+DateTimeFromatter syntaxes:#
```
Comment (mattyb149) [inline nifi-docs/src/main/asciidoc/expression-language-guide.adoc:2243]: Typo, should be `DateTimeFormatter`

File: nifi-docs/src/main/asciidoc/expression-language-guide.adoc [hunk 2]
```diff
@@ -2233,8 +2233,40 @@ In order to run the correct method, the parameter types must be correct. The Exp
...
+
+*Examples*: If the attribute "time" has the value "1647884009479", timeIsoOffset has the value "2022-12-03T10:15:30+01:00",
+timeIsoInstant has the value "2022-12-03T10:15:30Z" and "timeRFC1123" has the value "Thu, 01 Dec 2022 16:00:00 GMT",
+then the following Expressions will yield the following results:
+
+.format Examples
+|============================================================================
+| Expression | Value
+| `${time:formatInstant("yyyy/MM/dd HH:mm:ss.SSS\'Z'", "GMT")}` | `2022/03/21 17:33:29.479Z`
+| `${timeIsoOffset:format("yyyy/MM/dd HH:mm:ss.SSS\'Z'", "Asia/Tokyo")}` | `2022/12/03 18:15:30.000Z`
```
Comment (mattyb149) [inline nifi-docs/src/main/asciidoc/expression-language-guide.adoc:2266]: Should these be calls to `formatInstant`?

File: nifi-docs/src/main/asciidoc/expression-language-guide.adoc [hunk 3]
```diff
@@ -2264,6 +2296,33 @@ chaining together the two functions: `${date:toDate('MM-dd-yyyy'):format('yyyy/M
...
+
+*Arguments*:
+
+- [.argName]#_format_# : [.argDesc]#The current format to use when parsing the Subject, in the Java DateTimeFormatter syntax.#
+- [.argName]#_time zone_# : [.argDesc]#Optional argument that specifies the time zone to use when parsing the Subject, in the Java TimeZone syntax.#
+
+
+*Return Type*: [.returnType]#Instant#
+
+*Examples*: If the attribute "time" has the value "2014/12/31 15:36:03.264Z" then the expression `${time:toDate("yyyy/MM/dd HH:mm:ss.SSS'Z'", "GMT")}` will result in a Date data type for
```
Comment (mattyb149) [inline nifi-docs/src/main/asciidoc/expression-language-guide.adoc:2317]: Should this be calling `toInstant()`? and doesn't it return an Instant object (or a String made from one)?

File: nifi-docs/src/main/asciidoc/expression-language-guide.adoc [hunk 4]
```diff
@@ -2264,6 +2296,33 @@ chaining together the two functions: `${date:toDate('MM-dd-yyyy'):format('yyyy/M
...
+15:36:03.264 GMT on December 31, 2014.
+
+Often, this function is used in conjunction with the <<formatInstant>> function to change the format of a date/time. For example,
+if the attribute "instant" has the value "12-24-2014 12:06:59" and we want to change the format to "2014/12/24", we can do so by
+chaining together the two functions: `${date:toDate('MM-dd-yyyy HH:mm:ss'):format('yyyy/MM/dd')}`.
```
Comment (mattyb149) [inline nifi-docs/src/main/asciidoc/expression-language-guide.adoc:2322]: Should this be using `toInstant()` and/or `formatInstant()`?

File: nifi-docs/src/main/asciidoc/expression-language-guide.adoc (hunk 3 above)
Comment (Lehel44) [reply nifi-docs/src/main/asciidoc/expression-language-guide.adoc:2317]: That's right!

File: nifi-docs/src/main/asciidoc/expression-language-guide.adoc [hunk 5]
```diff
@@ -2299,30 +2299,58 @@ chaining together the two functions: `${date:toDate('MM-dd-yyyy'):format('yyyy/M
...
+*Description*: [.description]#Converts an Instant Subject into a Number with nanosecond precision#
+
+*Subject Type*: [.subject]#Instant#
+
+*Arguments*: No arguments
+
+*Return Type*: [.returnType]#Number#
 
+*Examples*: If the "dateTime" attribute has the value of "2022/03/18 10:22:27.678234567", then the Expression
+`${dateTime:toInstant('yyyy/MM/dd HH:mm:ss.SSSSSS', 'America/New_York'):toNanos()}` converts the Instant value of
```
Comment (mattyb149) [inline nifi-docs/src/main/asciidoc/expression-language-guide.adoc:2352]: I think the `S`s should be `n`s and there should be 9?

Comment (Lehel44) [reply nifi-docs/src/main/asciidoc/expression-language-guide.adoc]: According to the DateTimeFormatter [documentation](https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html) _S_ means fraction-of-second. _n_ works also. I changed it to 9 S from 6.

File: nifi-commons/nifi-expression-language/src/main/java/org/apache/nifi/attribute/expression/language/compile/ExpressionCompiler.java [hunk 6]
```diff
@@ -559,6 +572,18 @@ private DateEvaluator toDateEvaluator(final Evaluator<?> evaluator, final String
         return new DateCastEvaluator(evaluator);
     }
 
+    private InstantEvaluator toInstantEvaluator(final Evaluator<?> evaluator) {
+        return toInstantEvaluator(evaluator, null);
+    }
+
+    private InstantEvaluator toInstantEvaluator(final Evaluator<?> evaluator, final String location) {
```
[the remark this answers is not in the document]
Comment (Lehel44) [reply nifi-commons/nifi-expression-language/src/main/java/org/apache/nifi/attribute/expression/language/compile/ExpressionCompiler.java:579]: I followed the toDateEvaluator pattern and that doesn't use the location.

**Pull request closed without merging**